### ISSUE TITLE
Valid transactions don't seem to be getting mined

### ISSUE THREAD

**Issue by MarvinJanssen** [stacks-network/stacks-blockchain#2772]
Posting by request of @wileyj.

Curiously, an empty block was mined at #21,832 even though there were plenty of transactions waiting in the mempool. I only noticed this by accident. 

https://explorer.stacks.co/block/0xd7805cfeefc36cec46c19113a82a44912818e1b5fb2a22f80b25599457687e0b?chain=mainnet

**Comment by kantai** [comment]
It's hard to diagnose why a miner mined an empty block without actually having access to the node in question.

One way to obtain some mining diagnostics is to run a "mock miner" (set the `mock_mining` config setting to `True`) and set the miner key to something random (that doesn't have BTC). This will configure your node to _attempt_ to mine blocks, but always fail. This way, you'd be able to see when the miner is assembling a block what transactions it tries to include, etc.

**Comment by wileyj** [comment]
bump. we've been seeing a few more instances of this

**Comment by diwakergupta** [comment]
If there are specific txids that exhibit this behavior, that would help tremendously - thanks!

**Comment by gregorycoppola** [comment]
Looks like all the errors I am seeing in the mempool right now, from running a mock miner, are from bad nonce issues, e.g.:

```
WARN [1628896124.771645] [src/chainstate/stacks/miner.rs:1538] [relayer] Transaction skipped: InvalidStacksTransaction on tx a854016a91cd1c29e4a16f3fad1a757ebf93b47ade4bf61dce65a263ca3bd2ef, msg: Bad nonce: origin account SP1ESRDHP6T60Q25QY8TY7NZ18439DB69ZPFX40S3 nonce of tx a854016a91cd1c29e4a16f3fad1a757ebf93b47ade4bf61dce65a263ca3bd2ef is 8 (expected 11)
```

**Comment by diwakergupta** [comment]
Here's an example (via @friedger ): https://explorer.stacks.co/txid/0x4c7adc993be031d1ae61b9e26e9e0c7f881f871e24d852825c865f1fbd673922

Tx ID `0x4c7adc993be031d1ae61b9e26e9e0c7f881f871e24d852825c865f1fbd673922`, tx nonce 124, account nonce 124.

Mock miner (w/ @gregorycoppola 's changes) says `Transaction added: Block mined successfully 4c7adc993be031d1ae61b9e26e9e0c7f881f871e24d852825c865f1fbd673922.` but it's still pending.

**Comment by friedger** [comment]
The tx was generated by a script, they are quite big in terms of resources.

**Comment by friedger** [comment]
Here is another tx that sits in the mempool than it should I think
https://explorer.stacks.co/txid/0x095a9d00b09f05cc3dc0eb0c8997a473c292998b7b3ac01387cdccd8b8e84dce

**Comment by yknl** [comment]
Also seeing this behaviour with Xverse pool's Stacking transactions. At the moment several of the transactions have been pending for 4 blocks since being broadcasted. I've verified that the nonces are correct.

Some examples:
https://explorer.stacks.co/txid/0x44d43b6050f918e8df6eeae821089999b0810de33a6affd4c52f3d180de0da27
https://explorer.stacks.co/txid/0xc5986061cbb007e5f2e21c61b0ca24f7a90b8949ee15ed400b6794a2f0c324a8

**Comment by jcnelson** [comment]
The issue seems to be that the current mempool walk logic only considers the origin account nonce, not the transaction fee.  It prioritizes low-nonce transactions over high-nonce transactions, and prioritizes earlier transactions over later transactions.  This needs to be fixed -- the walk logic should order transactions by the origin addresses with the highest *cumulative* outstanding transaction fees in the mempool.

**Comment by jcnelson** [comment]
Fixed in #2823
